File: clients/cs/Game.cs
using System.Diagnostics;

namespace cs_snake
{
    public class Game
    {
        private static Game? sInstance;

        public static void Init()
        {
            Debug.Assert(sInstance == null);
            sInstance = new Game();
        }

        public static Game GetInstance()
        {
            Debug.Assert(sInstance != null);
            return sInstance;
        }

        public bool IsRunning;
        public Core Core;
        public Renderer Renderer;

        private Game()
        {
            IsRunning = true;
            Core = new Core(30, 30);
            Renderer = new Renderer("CS_SNAKE", Core, 20);
        }

        public void Run()
        {
            while (!Renderer.ShouldClose)
            {
                Renderer.Update();
                Core.Update();
                Renderer.Draw();
            }

            Renderer.Destroy();
            Core.Destroy();
        }


    }
}

File: clients/ts/src/game.ts
import { Core } from "./core";
import { Renderer } from "./renderer";

export class Game {
  core: Core;
  renderer: Renderer;
  constructor() {
    this.core = new Core(30, 30);
    this.renderer = new Renderer(this.core, 20);
  }

  run() {
    const runFrame = () => {
      this.renderer.update();
      this.renderer.draw();
      requestAnimationFrame(runFrame);
    };
    runFrame();
    // Cannot clean up because technically will never reach end of lifecycle unless the page is closed?
    // Maybe implement clean up if there is a restart button or move on to a different page
  }
}
